hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/HoodieSparkIndexClient.java

```diff
@@ -133,51 +130,43 @@ private void createRecordIndex(HoodieTableMetaClient metaClient, String userInde
-    String fullIndexName = indexType.equals(PARTITION_NAME_SECONDARY_INDEX)
-        ? PARTITION_NAME_SECONDARY_INDEX_PREFIX + userIndexName
-        : PARTITION_NAME_EXPRESSION_INDEX_PREFIX + userIndexName;
-    if (indexExists(metaClient, fullIndexName)) {
-      throw new HoodieMetadataIndexException("Index already exists: " + userIndexName);
-    }
-    checkArgument(columns.size() == 1, "Only one column can be indexed for functional or secondary index.");
-
-    if (!isEligibleForIndexing(metaClient, indexType, tableProperties, columns)) {
-      throw new HoodieMetadataIndexException("Not eligible for indexing: " + indexType + ", indexName: " + userIndexName);
-    }
-
-    HoodieIndexDefinition indexDefinition = new HoodieIndexDefinition(fullIndexName, indexType, options.getOrDefault(EXPRESSION_OPTION, IDENTITY_FUNCTION),
-        new ArrayList<>(columns.keySet()), options);
+    HoodieIndexDefinition indexDefinition = HoodieIndexUtils.getSecondaryOrExpressionIndexDefinition(metaClient, userIndexName, indexType, columns, options, tableProperties);
     if (!metaClient.getTableConfig().getRelativeIndexDefinitionPath().isPresent()
         || !metaClient.getIndexMetadata().isPresent()
-        || !metaClient.getIndexMetadata().get().getIndexDefinitions().containsKey(fullIndexName)) {
+        || !metaClient.getIndexMetadata().get().getIndexDefinitions().containsKey(indexDefinition.getIndexName())) {
       LOG.info("Index definition is not present. Registering the index first");
       register(metaClient, indexDefinition);
     }
 
     ValidationUtils.checkState(metaClient.getIndexMetadata().isPresent(), "Index definition is not present");
 
-    LOG.info("Creating index {} of using {}", fullIndexName, indexType);
+    LOG.info("Creating index {} of using {}", indexDefinition.getIndexName(), indexType);
     Option<HoodieIndexDefinition> expressionIndexDefinitionOpt = Option.ofNullable(indexDefinition);
     try (SparkRDDWriteClient writeClient = getWriteClient(metaClient, expressionIndexDefinitionOpt, Option.of(indexType))) {
       MetadataPartitionType partitionType = indexType.equals(PARTITION_NAME_SECONDARY_INDEX) ? MetadataPartitionType.SECONDARY_INDEX : MetadataPartitionType.EXPRESSION_INDEX;
       // generate index plan
-      Option<String> indexInstantTime = doSchedule(writeClient, metaClient, fullIndexName, partitionType);
+      Option<String> indexInstantTime = doSchedule(writeClient, metaClient, indexDefinition.getIndexName(), partitionType);
       if (indexInstantTime.isPresent()) {
         // build index
         writeClient.index(indexInstantTime.get());
       } else {
         throw new HoodieMetadataIndexException("Scheduling of index action did not return any instant.");
       }
     } catch (Throwable t) {
-      drop(metaClient, fullIndexName, Option.ofNullable(indexDefinition));
+      drop(metaClient, indexDefinition.getIndexName(), Option.ofNullable(indexDefinition));
```

Comment:
can we add a warn log here that indexing failed and so we are dropping the index.